activerecord/test/cases/relations_test.rb

```diff
@@ -836,6 +836,35 @@ def test_find_by_with_delegated_ar_object
     assert_equal author, Author.find_by(id: [decorator.new(author)])
   end
 
+  class CustomClassWithId
+    attr_reader :id, :name
+    def initialize(id, name); @id = id; @name = name end
+    def to_s; "#{@name}:#{@id}" end
+  end
+
+  class MySerializer
+    def self.load(object)
+      return unless object.is_a? String
+
+      name, id = object.split(":")
+
+      CustomClassWithId.new(id, name)
+    end
+    def self.dump(object); object.to_s end
+  end
```

Comment:
Maybe we can make the test case clearer by making the classes a bit more descriptive.
We can also ignore the unused `name` and the String guard clause.

```suggestion
  class ValueWithId
    def initialize(value); @value = value end
    def id; @value.split(':').first end
    def to_s; @value end
  end

  class ValueWithIdSerializer
    def self.load(value)
      ValueWithId.new(value)
    end
    def self.dump(object); object.to_s end
  end
```

Reply:
I'm afraid we don't have a passing test with the above changes. The reason is that the `ValueWithId` needs to be an object with 2 parts: an id integer and any other attribute of any other type for the assertion to make sense.
I think an even better naming would be `ObjectWithId`.